File: python/activities/tools.py
"""The EXECUTE TOOLS step (slide 28, primitive 03): tool calls as an Activity.

One dispatch over the tools → the plain SQL functions in db.py.
Sync code (psycopg) — the worker runs it on a thread pool.

Business errors (unknown destination, bad IDs) return as model-visible tool
results so the agent can explain them; they are NOT activity failures.
Infrastructure errors (DB down) raise → Temporal retries the activity.
"""

import json

from temporalio import activity
from temporalio.exceptions import ApplicationError

from . import db
from models.types import ToolRequest


def _settle_booking(account_key: str, items: list[dict], summary: str) -> dict:
    """Book the itinerary with one business rule: you can't book the same flight
    or hotel twice.

    That's the UNRECOVERABLE failure — a non-retryable BookingDeclined that
    Temporal does NOT retry (contrast the LLM kill-switch, which IS retried).
    The workflow surfaces it to the traveller and the conversation continues.
    """
    already = db.items_already_booked(account_key, items)
    if already:
        names = ", ".join(f'"{a["title"]}"' for a in already)
        raise ApplicationError(
            f"Booking declined — this trip already includes {names}.",
            type="BookingDeclined",
            non_retryable=True,
        )
    return db.record_booking(account_key, items, summary)


@activity.defn
def execute_tool(req: ToolRequest) -> str:
    name, args = req.call.name, req.call.args
    try:
        if name == "search_events":
            result = db.search_events(args["destination"], args.get("month"))
        elif name == "search_destinations":
            result = db.search_destinations(args["query"])
        elif name == "get_destination_info":
            result = db.get_destination_info(args["destination"])
        elif name == "search_flights":
            result = db.search_flights(
                args["destination"], args.get("origin"), args.get("depart_date")
            )
        elif name == "search_hotels":
            result = db.search_hotels(args["destination"], args.get("max_price"))
        elif name == "search_attractions":
            result = db.search_attractions(args["destination"])
        elif name == "get_bookings":
            result = db.get_bookings(req.account_key)
        elif name == "get_booking_details":
            result = db.get_booking_details(args["booking_id"])
        elif name == "add_to_itinerary":
            # Just the lookup — the workflow owns itinerary STATE (durable,
            # conversation-scoped). Unknown refs are dropped so the model reacts.
            result = db.get_itinerary_items(args["items"])
        elif name == "book_trip":
            result = _settle_booking(req.account_key, args["items"], args.get("summary", ""))
        elif name == "create_invoice":
            # Runs only after the traveller confirms (the workflow gates it).
            result = db.record_invoice(req.account_key, args["amount"], args["flight_details"])
        else:
            raise ApplicationError(f"Unknown tool: {name}", non_retryable=True)
    except ValueError as e:
        return json.dumps({"error": str(e)})  # business error → the model handles it
    return json.dumps(result, default=str)

File: typescript/src/activities/tools.ts
// The EXECUTE TOOLS step: tool calls as an Activity. One dispatch over the tools
// → the plain SQL functions in db.ts. Ported from python/activities/tools.py.
//
// Business errors (unknown destination, bad IDs, empty itinerary) come back as
// model-visible tool results so the agent can explain them; they are NOT
// activity failures. Infrastructure errors (DB down) throw → Temporal retries.
// A duplicate booking is the one UNRECOVERABLE failure — a non-retryable
// BookingDeclined that Temporal does not retry.

import { ApplicationFailure } from '@temporalio/common';

import * as db from './db';
import type { ToolRequest } from '../types';

// Book the itinerary with one business rule: you can't book the same flight or
// hotel twice. That duplicate is the non-retryable BookingDeclined.
async function settleBooking(
  accountKey: string,
  items: { kind: string; ref_id: number; title?: string; price?: number }[],
  summary: string
): Promise<Record<string, unknown>> {
  const already = await db.itemsAlreadyBooked(accountKey, items);
  if (already.length) {
    const names = already.map((a) => `"${a.title}"`).join(', ');
    throw ApplicationFailure.create({
      message: `Booking declined — this trip already includes ${names}.`,
      type: 'BookingDeclined',
      nonRetryable: true,
    });
  }
  return db.recordBooking(accountKey, items, summary);
}

export async function executeTool(req: ToolRequest): Promise<string> {
  const { name, args } = req.call;
  let result: unknown;
  try {
    switch (name) {
      case 'search_events':
        result = await db.searchEvents(args.destination as string, args.month as string | undefined);
        break;
      case 'search_destinations':
        result = await db.searchDestinations(args.query as string);
        break;
      case 'get_destination_info':
        result = await db.getDestinationInfo(args.destination as string);
        break;
      case 'search_flights':
        result = await db.searchFlights(
          args.destination as string,
          args.origin as string | undefined,
          args.depart_date as string | undefined
        );
        break;
      case 'search_hotels':
        result = await db.searchHotels(args.destination as string, args.max_price as number | undefined);
        break;
      case 'search_attractions':
        result = await db.searchAttractions(args.destination as string);
        break;
      case 'get_bookings':
        result = await db.getBookings(req.account_key);
        break;
      case 'get_booking_details':
        result = await db.getBookingDetails(args.booking_id as number);
        break;
      case 'add_to_itinerary':
        // Just the lookup — the workflow owns itinerary STATE (durable,
        // conversation-scoped). Unknown refs are dropped so the model reacts.
        result = await db.getItineraryItems(args.items as { kind?: string; id?: number }[]);
        break;
      case 'book_trip':
        result = await settleBooking(
          req.account_key,
          args.items as { kind: string; ref_id: number; title?: string; price?: number }[],
          (args.summary as string) ?? ''
        );
        break;
      case 'create_invoice':
        // Runs only after the traveller confirms (the workflow gates it).
        result = await db.recordInvoice(
          req.account_key,
          args.amount as number,
          args.flight_details as string
        );
        break;
      default:
        throw ApplicationFailure.create({ message: `Unknown tool: ${name}`, nonRetryable: true });
    }
  } catch (e) {
    if (e instanceof db.BusinessError) {
      return JSON.stringify({ error: e.message }); // business error → the model handles it
    }
    throw e;
  }
  return JSON.stringify(result);
}
